src/app/fdctl/run/tiles/shred.c

```diff
@@ -0,0 +1,174 @@
+#define _GNU_SOURCE
+#include <linux/unistd.h>
+#include <sys/mman.h> /* For mmap, etc. */
+#include <string.h> /* for explicit_bzero */
+
+#include "tiles.h"
+#include "../../fdctl.h"
+#include "../run.h"
+
+#include "../../../../disco/fd_disco.h"
+#include "../../../../disco/shred/fd_shred_tile.h"
+#include "../../../../tango/ip/fd_ip.h"
+
+
+#define FD_SHRED_TAG (0x5117ed711eUL) /* SHRED TILE */
+
+
+
+static void
+init( fd_tile_args_t * args ) {
+  void * _ip = fd_wksp_alloc_laddr( fd_wksp_containing( args->wksp_pod[ 0 ] ), fd_ip_align(), fd_ip_footprint( 256UL, 256UL ), FD_SHRED_TAG );
+  fd_ip_t * ip = (void *)fd_ip_join( fd_ip_new( _ip, 256UL, 256UL ) );
+  if( FD_UNLIKELY( !ip ) ) FD_LOG_ERR(( "fd_ip_join failed" ));
+  args->lo_xsk = (void *)ip; /* FIXME: Hack */
+
+  /* calling fd_tempo_tick_per_ns requires nanosleep, it is cached with
+     a FD_ONCE */
+  fd_tempo_tick_per_ns( NULL );
+
+  const char * identity_key_path = fd_pod_query_cstr( args->wksp_pod[ 0 ], "identity_key_path", NULL );
+  if( FD_UNLIKELY( !identity_key_path ) ) FD_LOG_ERR(( "identity_key_path not found in tpod" ));
+
+  uchar const * key_page = load_key_into_protected_memory( identity_key_path );
+
+  args->xsk = (void *)key_page; /* FIXME: Fix this after topology merge */
+}
+
+static void
+run( fd_tile_args_t * tile_args ) {
+  uchar const * tile_pod        = tile_args->wksp_pod[ 0 ];
+  uchar const * bank_shred_pod  = tile_args->wksp_pod[ 1 ];
+  uchar const * shred_store_pod = tile_args->wksp_pod[ 2 ];
+  uchar const * net_shred_pod   = tile_args->wksp_pod[ 3 ];
+
+  fd_shred_tile_args_t args = {0};
+  args.fec_resolver_depth = fd_pod_query_ulong( tile_pod, "fec_resolver_depth", 0UL );
+  if( FD_UNLIKELY( !args.fec_resolver_depth ) ) FD_LOG_ERR(( "fec_resolver_depth missing from pod or 0" ));
+
+  args.fec_resolver_done_depth = fd_pod_query_ulong( tile_pod, "fec_resolver_done_depth", args.fec_resolver_depth*128UL );
+
+  ulong bank_cnt = fd_pod_query_ulong( bank_shred_pod, "cnt", 0UL );
+  if( FD_UNLIKELY( !bank_cnt ) ) FD_LOG_ERR(( "0 bank tiles" ));
+  if( FD_UNLIKELY( bank_cnt>MAX_BANK_CNT ) ) FD_LOG_ERR(( "Too many banks" ));
+  args.bank_cnt = bank_cnt;
+
+
+  args.src_mac  = fd_pod_query_buf( tile_pod, "src_mac_addr", NULL );
+  if( FD_UNLIKELY( !args.src_mac ) ) FD_LOG_ERR(( "src_mac_addr missing from pod" ));
+  args.src_ip = fd_pod_query_uint( tile_pod, "ip_addr", 0 );
+  if( FD_UNLIKELY( !args.src_ip  ) ) FD_LOG_ERR(( "ip_addr missing from pod"      ));
+
+  ulong * _shred_version = fd_wksp_pod_map( tile_pod, "shred_version" );
+  if( FD_UNLIKELY( !_shred_version ) ) FD_LOG_ERR(( "shred_version missing from pod" ));
+  ulong shred_version = FD_VOLATILE_CONST( *_shred_version );
+  if( shred_version == 0UL ) {
+    FD_LOG_NOTICE(( "Waiting for shred version to be determined via gossip.  "
+          "To avoid this, specify the shred version in the configuration TOML file." ));
```

Comment:
Kind of a suspect comment? I don't think we want to recommend to user setting this value in the config, they will likely get it wrong.